### PR TITLE
New feature support / quality improvements

### PR DESCRIPTION
**Pull request by hassount**
- Trellis
- Multi-reference pictures support
- Chroma search
- Tiles enc dec mismatch fix / tiles bitrate overhead fix
- Enhanced Adaptive Depth partitioning algorithm
- Cabac Update
- New M0-M8 Preset tuning

### CODE REVIEW

File: Source/Lib/Common/ASM_AVX2/EbPictureOperators_Intrinsic_AVX2.c [hunk 1]
```diff
@@ -2831,3 +2832,213 @@ void ResidualKernel_avx2(
...
+{
+    int32_t row_count = area_height;
+    __m256i sum = _mm256_setzero_si256();
+    __m128i sum_L, sum_H, s;
+
+    (void)area_width;
+
+    do
+    {
+        const __m128i in0 = _mm_cvtsi32_si128(*(uint32_t *)(input + 0 * input_stride));
```
Comment (joelsole) [inline Source/Lib/Common/ASM_AVX2/EbPictureOperators_Intrinsic_AVX2.c]: Style - several instances. Replace by

 do {

File: Source/Lib/Common/Codec/EbDefinitions.h [hunk 2]
```diff
@@ -3254,13 +3428,385 @@ static const uint32_t raster_scan_cu_parent_index[CU_MAX_COUNT] =
...
+    /*78 */3,
+    /*79 */3,
+    /*80 */3,
+    /*81 */3,
+    /*82 */3,
+    /*83 */3,
+    /*84 */3,
+};
 
 #if SCENE_CONTENT_SETTINGS
```
[the remark this answers is not in the document]
Comment (hassount) [reply Source/Lib/Common/Codec/EbDefinitions.h]: typo, it should be SCREEN_CONTENT_SETTINGS, Fixed

File: Source/Lib/Common/Codec/EbDefinitions.h [hunk 3]
```diff
@@ -34,105 +34,224 @@
 #ifdef __cplusplus
 extern "C" {
 #endif
-    
-#define QUICK_ME_CLEANUP        1
-#define SCENE_CONTENT_SETTINGS  1
+#define WIKI_SETTINGS          0
```
[the remark this answers is not in the document]
Comment (hassount) [reply Source/Lib/Common/Codec/EbDefinitions.h]: Agreed, Fixed

File: Source/Lib/Common/ASM_AVX2/EbPictureOperators_Intrinsic_AVX2.c (hunk 1 above)
Comment (hassount) [reply Source/Lib/Common/ASM_AVX2/EbPictureOperators_Intrinsic_AVX2.c]: Fixed

**Comment by hassount** [comment]
> Please, could you fix the warnings on linux and macos?

Done, the rest will be fixed by PR 234

File: Source/Lib/Common/Codec/EbEncDecProcess.c [hunk 4]
```diff
@@ -1573,6 +1823,48 @@ void* enc_dec_kernel(void *input_ptr)
...
+                    {
+                        MdRateEstimationContext* md_rate_estimation_array = sequence_control_set_ptr->encode_context_ptr->md_rate_estimation_array;
+                        md_rate_estimation_array += picture_control_set_ptr->slice_type * TOTAL_NUMBER_OF_QP_VALUES + context_ptr->md_context->qp;
+
+                        //this is temp, copy all default tables
+                        picture_control_set_ptr->rate_est_array[sb_index] = *md_rate_estimation_array;
+#if CABAC_SERIAL
+                        if (sb_index == 0) {
+                            picture_control_set_ptr->ec_ctx_array[sb_index] = *picture_control_set_ptr->coeff_est_entropy_coder_ptr->fc;
+                        }
```
[the remark this answers is not in the document]
Comment (hassount) [reply Source/Lib/Common/Codec/EbEncDecProcess.c]: Done

File: Source/Lib/Common/Codec/EbEncDecProcess.c [hunk 5]
```diff
@@ -1573,6 +1823,48 @@ void* enc_dec_kernel(void *input_ptr)
                     context_ptr->sb_index = sb_index;
                     context_ptr->md_context->cu_use_ref_src_flag = (picture_control_set_ptr->parent_pcs_ptr->use_src_ref) && (picture_control_set_ptr->parent_pcs_ptr->edge_results_ptr[sb_index].edge_block_num == EB_FALSE || picture_control_set_ptr->parent_pcs_ptr->sb_flat_noise_array[sb_index]) ? EB_TRUE : EB_FALSE;
 
+#if CABAC_UP
+                    if (picture_control_set_ptr->update_cdf)
+                    {
```
[the remark this answers is not in the document]
Comment (hassount) [reply Source/Lib/Common/Codec/EbEncDecProcess.c]: Done

File: Source/Lib/Common/Codec/EbEntropyCoding.c [hunk 6]
```diff
@@ -2404,7 +2772,125 @@ int32_t av1_get_pred_context_single_ref_p6(const MacroBlockD *xd) {
...
+                const int bit2 = mbmi->ref_frame[0] == GOLDEN_FRAME;
+                WRITE_REF_BIT(bit2, comp_ref_p2);
+            }
+
+            const int bit_bwd = mbmi->ref_frame[1] == ALTREF_FRAME;
+            WRITE_REF_BIT(bit_bwd, comp_bwdref_p);
+
+            if (!bit_bwd) {
+                WRITE_REF_BIT(mbmi->ref_frame[1] == ALTREF2_FRAME, comp_bwdref_p1);
+            }
```
[the remark this answers is not in the document]
Comment (hassount) [reply Source/Lib/Common/Codec/EbEntropyCoding.c]: Done

File: Source/Lib/Common/Codec/EbEntropyCoding.c [hunk 7]
```diff
@@ -2404,7 +2772,125 @@ int32_t av1_get_pred_context_single_ref_p6(const MacroBlockD *xd) {
...
+static void write_ref_frames(
+    FRAME_CONTEXT               *frame_context,
+    PictureParentControlSet   *pcs_ptr,
+    const MacroBlockD           *xd,
+    AomWriter                  *w) {
+
+    const MbModeInfo *const mbmi = &xd->mi[0]->mbmi;
+    const int is_compound = has_second_ref(mbmi);
+    UNUSED(frame_context);
+#if 0
```
[the remark this answers is not in the document]
Comment (hassount) [reply Source/Lib/Common/Codec/EbEntropyCoding.c]: Done

File: Source/Lib/Common/Codec/EbEntropyCoding.c [hunk 8]
```diff
@@ -2043,7 +2076,305 @@ int32_t av1_get_reference_mode_context(
...
+            ref_counts[above_mbmi->ref_frame[1]]++;
+        }
+    }
+
+    // Left neighbor
+    if (left_in_image && is_inter_block(left_mbmi)) {
+        ref_counts[left_mbmi->ref_frame[0]]++;
+        if (has_second_ref(left_mbmi)) {
+            ref_counts[left_mbmi->ref_frame[1]]++;
+        }
```
[the remark this answers is not in the document]
Comment (hassount) [reply Source/Lib/Common/Codec/EbEntropyCoding.c]: Done

File: Source/Lib/Common/Codec/EbFullLoop.c [hunk 9]
```diff
@@ -488,42 +488,1402 @@ void av1_quantize_b_facade_II(
...
+    const ScanOrder *const scan_order = &av1_scan_orders[tx_size][tx_type];
+#endif
+#if 0
+    const ScanOrder *scan_order = get_scan(tx_size, tx_type);
+#endif
+#if !DEBUG_TRELLIS
+    const int16_t *scan = scan_order->scan;
+#endif
+    const int shift = av1_get_tx_scale(tx_size);
+#if 0
```
[the remark this answers is not in the document]
Comment (hassount) [reply Source/Lib/Common/Codec/EbFullLoop.c]: Done

File: Source/Lib/Common/Codec/EbMdRateEstimation.c [hunk 10]
```diff
@@ -67,7 +67,228 @@ void av1_get_syntax_rate_from_cdf(
...
+
+#if CABAC_UP2
+    //if (cm->skip_mode_flag) { // NM - Hardcoded to true
+    for (i = 0; i < SKIP_CONTEXTS; ++i) {
+        av1_get_syntax_rate_from_cdf(md_rate_estimation_array->skipModeFacBits[i], fc->skip_mode_cdfs[i], NULL);
+    }
+    //}
+#endif
+
+#if 0
```
[the remark this answers is not in the document]
Comment (hassount) [reply Source/Lib/Common/Codec/EbMdRateEstimation.c]: Done

File: Source/Lib/Common/Codec/EbFullLoop.c [hunk 11]
```diff
@@ -488,42 +488,1402 @@ void av1_quantize_b_facade_II(
...
+    EbBool                  is_inter,
+    uint32_t                bit_increment,
+    int                     plane)
+
+{
+    (void)stride;
+    (void)n_coeffs;
+    (void)sc;
+    (void)qparam;
+#if DEBUG_TRELLIS
```
[the remark this answers is not in the document]
Comment (hassount) [reply Source/Lib/Common/Codec/EbFullLoop.c]: Done, unit test will be done in a separate PR

**Comment by EwoutH** [comment]
I measured a [significant slowdown](https://docs.google.com/spreadsheets/d/1p3PJQMkyhIrXEL6MBBQiwTnWmGyHY_zuw7_XAD0def4/edit#gid=396997821) between April 8th and today, most likely caused by this PR. Was this expected (in exchange for higher quality), or a regression? 

@kylophone The AWCY run seemed to have stopped or failed, could you resubmit a run?

| Speed (fps) | 1bd31d3  | bef6750   |        |
|-------------|---------:|----------:|-------:|
| -enc-mode 8 |   35,461 |    24,704 | -30,3% |
| -enc-mode 7 |   23,424 |    14,471 | -38,2% |
| -enc-mode 6 |   15,544 |    10,671 | -31,3% |
| -enc-mode 5 |   11,580 |     5,016 | -56,7% |
| -enc-mode 4 |    5,396 |     2,307 | -57,2% |
| -enc-mode 3 |    2,419 |     1,708 | -29,4% |
| -enc-mode 2 |    1,416 |     0,680 | -52,0% |
| -enc-mode 1 |    0,988 |     0,337 | -65,9% |

**Comment by EwoutH** [comment]
Retested on a fresh system, results are now mixed.

| Speed (fps) | 1bd31d3  | bef6750   |        |
|-------------|---------:|----------:|-------:|
| -enc-mode 8 |   35,461 |    53,860 |  51,9% |
| -enc-mode 7 |   23,424 |    25,097 |   7,1% |
| -enc-mode 6 |   15,544 |    18,464 |  18,8% |
| -enc-mode 5 |   11,580 |     9,536 | -17,7% |
| -enc-mode 4 |    5,396 |     4,387 | -18,7% |
| -enc-mode 3 |    2,419 |     3,260 |  34,8% |
| -enc-mode 2 |    1,416 |     0,680 | -52,0% |
| -enc-mode 1 |    0,988 |     0,587 | -40,6% |
| -enc-mode 0 |    0,536 |     0,194 | -63,8% |

File: Docs/svt-av1_encoder_user_guide.md [hunk 12]
```diff
@@ -137,6 +137,7 @@ The encoder parameters present in the Sample.cfg file are listed in this table b
 | **IntraPeriod** | -intra-period | [-2 - 255] | -2 | Distance Between Intra Frame inserted. -1 denotes no intra update. -2 denotes auto. |
 | **IntraRefreshType** | -irefresh-type | [1 – 2] | 1 | 1: CRA (Open GOP)2: IDR (Closed GOP) |
 | **QP** | -q | [0 - 63] | 50 | Quantization parameter used when RateControl is set to 0 |
+| **RateControlMode** | -rc | [0 - 3] | 0 | 0 = CQP , 1 = ABR , 2 = ABR , 3 = CVBR |
```
Comment (HolyWu) [inline Docs/svt-av1_encoder_user_guide.md]: 1 = ABR , 2 = ABR ?

Comment (HolyWu) [inline Docs/svt-av1_encoder_user_guide.md]: @hassount
2 should be VBR.

**Pull request merged**